### ISSUE TITLE
Binding to new GameObject not putting the GameObject in the right scene

### ISSUE THREAD

**Issue by Soren025** [ssannandeji/Zenject-2019#263]
Here is my scenario.

I have two scenes, `Scene A` and `Scene B`.

Both have a `SceneContext`.

`Scene A` is currently the active scene.

`Scene B`'s `SceneContext` contains the following installer with `Parent New Objects Under Root` set to false.
```csharp
Container.Bind<Transform>()
    .FromNewComponentOnNewGameObject()
    .AsTransient()
    .NonLazy();
```

What you will find is that the new `GameObject` created by `Scene B`'s `SceneContext` ends up in `Scene A` because Unity places new `GameObject`s inside the active scene.

The solution for this would be to set the `SceneContext`'s `Scene` to be active right before any creation or instantiation of a `GameObject` that isn't parented to the `SceneContext`. If you are unaware you can get the `SceneContext`'s scene like this `sceneContext.gameobject.scene`.

My understanding is that the "active scene" feature is only used to determine where new `GameObject`s end up, so I don't believe this change would effect much.

**Comment by Soren025** [comment]
I just discovered a problem with my suggested solution. During the initial calls to `Awake()` and `OnEnable()` within a scene first loading up, you are unable to set it to be the active scene. It is only within `Start()` am I able to set the scene to be active.

This seems silly on unity's part as it makes it impossible for objects within a loading scene from creating new objects within the scene.

**Comment by Soren025** [comment]
Seems like if you parent it to an object, within the `SceneContext`'s scene and immediately set the parent to null, the created `GameObject` will end up in the intended scene.

**Comment by svermeulen** [comment]
This is a bug that I think we can fix.

The new game object is being created through Zenject, and so we should be able to figure out what scene it should be placed in, since we have the container, which has a reference to SceneContext.

So maybe the solution would be, when the new game object does not have an explicit parent, we explicitly set it to be inside the right scene.

**Comment by Soren025** [comment]
Indeed what I was going for with my suggested solution.

Also looking at my last comment, I should probably share what I did as code (rather than description) to bypass the fact that you can't set the the scene to be active during the initial `Awake()` call on a scene loading.

```csharp
SceneContext sceneContext;

GameObject gameObject = new GameObject();

// I have found that doing the following will place it in the same scene of the `SceneContext`
// regardless of what scene it was created in.
gameObject.transform.parent = sceneContext.transform;
gameObject.transform.parent = null;
```

**Comment by svermeulen** [comment]
Ah yes I see what you mean now.  Yeah that sounds like a good solution.   Maybe we would just need to change DiContainer.CreateAndParentPrefab to do this:

```
var gameObj = (GameObject)GameObject.Instantiate(
    prefabAsGameObject, GetTransformGroup(gameObjectBindInfo, context), false);

if (gameObj.transform.parent == null)
{
    gameObj.transform.parent = Resolve<Context>().transform;
    gameObj.transform.parent = null;
}
```

**Comment by Soren025** [comment]
Yes, that will do. But you should also do it for normal game object creation such as methods like `FromNewComponentOnNewGameObject()`.

However an interesting development I found out. When you shared your code snippet using `Context` I wondered how this would work out for `ProjectContext`. Executing the following code...

```csharp 
GameObject a = new GameObject("A");
DontDestroyOnLoad(a);

GameObject b = new GameObject("B");
b.transform.parent = a.transform;
b.transform.parent = null;
```

I discovered that if you parent a `GameObject` to an object set `DontDestroyOnLoad()` and unparent it. It will still maintain its state of `DontDestroyOnLoad()`. Here is a screenshot of the result,

![Test Result](http://puu.sh/vPhwl/2a1ad4c0aa.png)

**Comment by svermeulen** [comment]
Oh interesting.  That's actually a pretty nice result, since that's generally what you want to happen if you are instantiating something inside project context

**Comment by Nezz** [comment]
Are there any plans to fix this?

**Comment by svermeulen** [comment]
This is fixed on develop branch for version 6.0.  Might also be included in the current public release too.
